src/Components/Server/src/CircuitDisconnectMiddleware.cs

```diff
@@ -39,65 +39,66 @@ public async Task Invoke(HttpContext context)
                 return;
             }
 
-            var (hasCircuitId, circuitId) = await TryGetCircuitIdAsync(context);
-            if (!hasCircuitId)
+            var circuitSecret = await GetCircuitSecretAsync(context);
+            if (circuitSecret is null)
             {
                 context.Response.StatusCode = StatusCodes.Status400BadRequest;
                 return;
             }
 
-            await TerminateCircuitGracefully(circuitId);
+            await TerminateCircuitGracefully(circuitSecret.Value);
 
             context.Response.StatusCode = StatusCodes.Status200OK;
         }
 
-        private async Task<(bool, string)> TryGetCircuitIdAsync(HttpContext context)
+        private async Task<CircuitSecret?> GetCircuitSecretAsync(HttpContext context)
         {
             try
             {
                 if (!context.Request.HasFormContentType)
                 {
-                    return (false, null);
+                    return default;
                 }
 
                 var form = await context.Request.ReadFormAsync();
-                if (!form.TryGetValue(CircuitIdKey, out var circuitId) || !CircuitIdFactory.ValidateCircuitId(circuitId))
+                if (!form.TryGetValue(CircuitIdKey, out var text))
                 {
-                    return (false, null);
+                    return default;
                 }
 
-                return (true, circuitId);
+                var circuitSecret = new CircuitSecret(text);
+                if (!CircuitSecretFactory.ValidateCircuitSecret(circuitSecret))
+                {
+                    return default;
+                }
+
+                return circuitSecret;
             }
             catch
             {
-                return (false, null);
+                return default;
             }
         }
 
-        private async Task TerminateCircuitGracefully(string circuitId)
+        private async Task TerminateCircuitGracefully(CircuitSecret circuitSecret)
         {
-            try
-            {
-                await Registry.TerminateAsync(circuitId);
-                Log.CircuitTerminatedGracefully(Logger, circuitId);
-            }
-            catch (Exception e)
-            {
-                Log.UnhandledExceptionInCircuit(Logger, circuitId, e);
-            }
+            // We don't expect TerminateAsync to throw.
+            Log.CircuitTerminatingGracefully(Logger, circuitSecret);
```

Comment:
Should we not log the circuitId instead of the secret? (I guess we don't have access to the ID?). 

This is where hashing comes in handy. You can compute the ID directly from the secret and if you ever log the secret you can hash it to get the ID and correlate it with the rest of the logs you have for the circuit.

Right now a circuit will have a bunch of messages logged with the ID and we will not able to correlate graceful terminations for that circuit, and we could keep things as Debug here.

Reply:
Agreed.